plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+
+    genForall = do
+      x <- genMaybeFreshTyName "a"
+      k <- liftGen arbitrary
+      fmap (TyForall () x k) $ onSize (subtract 1) $ bindTyName x k $ genType Star
+
+    genLam k1 k2 = do
+        x <- genMaybeFreshTyName "a"
+        fmap (TyLam () x k1) $ onSize (subtract 1) $ bindTyName x k1 (genType k2)
+
+    genApp = do
+      k' <- liftGen arbitrary
+      sizeSplit_ 1 7 (genType $ KindArrow () k' k) (genType k') $ TyApp ()
+
+-- | Generate a closed type at a given kind
+genClosedType_ :: Kind () -> Gen (Type TyName DefaultUni ())
+genClosedType_ = genTypeWithCtx mempty
+
+-- | Generate a well-kinded term in a given context
+genTypeWithCtx :: Map TyName (Kind ()) -> Kind () -> Gen (Type TyName DefaultUni ())
+genTypeWithCtx ctx k = runGenTm $ local (\ e -> e { geTypes = ctx }) (genType k)
+
+-- | Generate a well-kinded term in a given context
+genKindAndTypeWithCtx :: Map TyName (Kind ()) -> Gen (Kind(), Type TyName DefaultUni ())
+genKindAndTypeWithCtx ctx = do
+  k <- arbitrary
+  runGenTm $ local (\ e -> e { geTypes = ctx }) ((k,) <$> genType k)
+
+-- CODE REVIEW: does this exist anywhere??
+-- | `substClosedType x sub ty` substitutes the closed type `sub` for `x` in `ty`
+substClosedType :: TyName -> Type TyName DefaultUni () -> Type TyName DefaultUni () -> Type TyName DefaultUni ()
+substClosedType x sub ty =
+  case ty of
+    TyVar _ y
+      | x == y    -> sub
+      | otherwise -> ty
+    TyFun _ a b   -> TyFun () (substClosedType x sub a) (substClosedType x sub b)
+    TyApp _ a b   -> TyApp () (substClosedType x sub a) (substClosedType x sub b)
+    TyLam _ y k b
+      | x == y    -> ty
+      | otherwise -> TyLam () y k $ substClosedType x sub b
+    TyForall _ y k b
+      | x == y    -> ty
+      | otherwise -> TyForall () y k $ substClosedType x sub b
+    TyBuiltin{}   -> ty
+    TyIFix{}      -> ty
+
+-- CODE REVIEW: does this exist anywhere?
+builtinKind :: SomeTypeIn DefaultUni -> Kind ()
+builtinKind (SomeTypeIn t) = case t of
+  DefaultUniProtoList -> Star :-> Star
+  DefaultUniProtoPair -> Star :-> Star :-> Star
+  DefaultUniApply f _ -> let _ :-> k = builtinKind (SomeTypeIn f) in k
+  _                   -> Star
+
+-- * Shrinking types and kinds
+
+-- | Shriking-order on kinds
+leKind :: Kind () -> Kind () -> Bool
+leKind k1 k2 = go (reverse $ args k1) (reverse $ args k2)
+  where
+    args Type{}            = []
+    args (KindArrow _ a b) = a : args b
+
+    go [] _                = True
+    go _ []                = False
+    go (k : ks) (k' : ks')
+      | leKind k k' = go ks ks'
+      | otherwise   = go (k : ks) ks'
+
+-- | Strict shrinking order on kinds
+ltKind :: Kind () -> Kind () -> Bool
+ltKind k k' = k /= k' && leKind k k'
+
+-- | Take a type in a context and a new target kind
+--   Precondition: new kind is smaller or equal to old kind of the type.
+--   TODO (later): also allow changing which context it's valid in
+fixKind :: HasCallStack
+        => Map TyName (Kind ())
```

Comment:
You know, I think it would help to either type-synonym or newtype your `Context` and `Substitution` types.